### ISSUE TITLE
NativeAOT compiler fails when it tries to emit a throw for BadImageFormatException

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/runtime#94107]
If the input for the AOT compiler is invalid in some sense, it should generate a throwing method body, but not fail the compilation.
In some cases this fails. An example is if we hit this code path:
https://github.com/dotnet/runtime/blob/77bd2ae57c67d9a7377305d4d1e169277c01daac/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs#L154

(happens when there's an unknown token type in the input)

This throws an internal exception in the compiler and attaches one argument to it.

Eventually this makes its way into the helper which generates the throwing method body where it will trigger an assert: https://github.com/dotnet/runtime/blob/77bd2ae57c67d9a7377305d4d1e169277c01daac/src/coreclr/tools/Common/TypeSystem/IL/Stubs/TypeSystemThrowingILEmitter.cs#L78

This is because the throw helper this will end up calling is:
https://github.com/dotnet/runtime/blob/77bd2ae57c67d9a7377305d4d1e169277c01daac/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs#L76

That takes only a single parameter, the exception string ID, but it doesn't take additional argument. But our internal exception has additional argument.

Eventually this ends up producing invalid IL - calls the helper with more arguments then it declares. This will eventually fail in the JIT and ends up failing the entire compilation with a cryptic error:
```
EXEC : error : One or more errors occurred. (Invalid IL or CLR metadata) []
  System.AggregateException: One or more errors occurred. (Invalid IL or CLR metadata)
   ---> Internal.TypeSystem.TypeSystemException+InvalidProgramException: Invalid IL or CLR metadata
     at Internal.TypeSystem.ThrowHelper.ThrowInvalidProgramException() + 0x24
  ....
```

A repro is:

`dotnet new console`

`app.csproj`:
```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net8.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
        <Nullable>enable</Nullable>
        <PublishAot>true</PublishAot>
        <!-- <PublishTrimmed>true</PublishTrimmed> -->
        <WarningsAsErrors>false</WarningsAsErrors>
        <NoWarn>NU1605;NU1603;NU1701</NoWarn>
        <RefAssemblyName>Microsoft.AspNetCore.Identity.EntityFrameworkCore</RefAssemblyName>
        <RefPackageVersion>3.1.10</RefPackageVersion>
        <RefPackageName></RefPackageName>
        <RefPackageName Condition="'$(PackageName)'==''">$(RefAssemblyName)</RefPackageName>
    </PropertyGroup>

    <ItemGroup>
        <TrimmerRootAssembly Include="$(RefAssemblyName)" />

        <PackageReference Include="$(RefPackageName)" Version="$(RefPackageVersion)" />
    </ItemGroup>

    <Target Name="ConfigureTrimming"
            BeforeTargets="PrepareForILLink">
        <ItemGroup>
        <ManagedAssemblyToLink Condition="'%(Filename)' == '$(RefAssemblyName)'">
            <TrimmerSingleWarn>false</TrimmerSingleWarn>
        </ManagedAssemblyToLink>
        </ItemGroup>
    </Target>
</Project>
```

`dotnet publish`

There's a side question about invalid IL in our library... but that's a separate problem (and probably not relevant as this is a really old version of the library)

**Comment by MichalStrehovsky** [comment]
This is https://github.com/dotnet/runtimelab/issues/1135. The crossgen2 folks don't care about the exceptions being translatable to a throwing stub. I keep pointing this out whenever I see a new exception being thrown in a pull request but I can't be on top of everything.

I distinctly remember this particularly problematic one because I commented on it (https://github.com/dotnet/runtime/pull/50437#discussion_r604755475) it was then fixed wrong (by introducing a constructor that takes a string instead of following what I suggested) and merged without further review because it was deemed a de minimis change by the PR author. I didn't comment on it because I was pretty upset about it and I wouldn't write anything nice. I knew it introduced a bug in the failure path. It also doesn't play with crossgen2's localization support at all so it's also a bug there.

It would be nice to introduce a mechanism that is less error prone here but I failed to come up with anything. (Also, it's hard to protect from someone just adding a new constructor to the exception class that immediately destroys everything.)

For ILC purposes, we need:

* To be able to print a message, potentially in the localization of the compiler (once we have localization; the ExceptionStringId system is set up to support that)
* To be able to take an exception object and create and IL stub from it that throws identical exception at runtime, using the localization of the corelib.

The existing system with the ExceptionStringIds is pretty fragile and hard to test. It would be great if we could replace this with something that is validated by the C# compiler at source code build time. I wasn't able to come up with anything.
